File: libmacrocenter/examples/basic.rs
use libmacrocenter::hotkey::HotkeyManager;
use libmacrocenter::input::InputSimulator;
#[allow(unused_imports)]
use libmacrocenter::types::{ActionMode, CoordinateMode, MouseButton, ScrollAxis};
use std::thread;
use std::time::Duration;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Input Simulation
    println!("=== Input Simulation ===");

    let mut _sim = InputSimulator::new()?;

    // Type some text
    // println!("Typing 'Hello from libmacrocenter!'...");
    // sim.type_text("Hello from libmacrocenter!")?;

    // Press a key
    // println!("Pressing Enter...");
    // sim.key_action("enter", ActionMode::Click)?;

    // Mouse actions
    // println!("Moving mouse to (500, 500)...");
    // sim.mouse_move(500, 500, CoordinateMode::Absolute)?;

    // println!("Clicking left mouse button...");
    // sim.mouse_click(MouseButton::Left, ActionMode::Click)?;

    // println!("Scrolling down...");
    // sim.scroll(ScrollAxis::Vertical, -3)?;

    // NOTE: The above calls are commented out to avoid actually
    // sending input when running the example. Uncomment them to
    // see them in action!

    // --- Hotkey Listening ---
    // println!("\n=== Hotkey Listening ===");

    // NOTE: global-hotkey requires an event loop on the current thread.
    // on Linux (X11), this example should work. On macOS/Windows,
    // you need a proper event loop running.
    let mut mgr = HotkeyManager::new()?;

    // register a hotkey
    let id = mgr.register("Ctrl+Shift+M")?;
    println!("Registered hotkey Ctrl+Shift+M with id: {id}");

    let f5_id = mgr.register("F5")?;
    println!("Registered hotkey F5 with id: {f5_id}");

    println!("\nListening for hotkeys... Press Ctrl+C to exit.");
    println!("Try pressing Ctrl+Shift+M or F5!\n");

    loop {
        if let Some(event) = mgr.poll_event() {
            let name = if event.id == id {
                "Ctrl+Shift+M"
            } else if event.id == f5_id {
                "F5"
            } else {
                "Unknown"
            };
            println!("Hotkey event: {name} -> {:?}", event.state);
        }

        thread::sleep(Duration::from_millis(10));
    }
}

File: libmacrocenter-node/examples/basic.ts
// build first: cd libmacrocenter-node && bun run build
// then run: node examples/basic.ts

import { InputSimulator, HotkeyManager } from "../libmacrocenter-node/index.js";

console.log("=== libmacrocenter Features Showcase ===");

// 1. Input Simulation
console.log("\n[1] Simulator capabilities:");
const sim = new InputSimulator();
// Type some text
// console.log("- Typing text:");
// sim.typeText("Hello from libmacrocenter!\nThis is typing simulation.\n");

// console.log("- Key actions (press, release, click):");
// sim.keyAction("shift", "press");
// sim.keyAction("a", "click"); // Types an uppercase A
// sim.keyAction("shift", "release");

// console.log("- Mouse movements:");
// sim.mouseMove(100, 100, "absolute"); // Moves to top left
// sim.mouseMove(50, 50, "relative");   // Moves 50px right and down

// console.log("- Mouse clicks:");
// sim.mouseClick("right", "click");

// console.log("- Scrolling:");
// sim.scroll("vertical", -5); // Scrolls down
// sim.scroll("horizontal", 2); // Scrolls right

//  2. Hotkey Management & Callbacks
console.log("\n[2] Hotkey Listening:");

const mgr = new HotkeyManager();

const shiftM = mgr.register("9");
// const f5 = mgr.register("Ctrl+Alt+Shift+F5");
const f5 = mgr.register("F5");

console.log(
  `Registered hotkeys: \n- Ctrl+Alt+Shift+M (id: ${shiftM})\n- Ctrl+Alt+Shift+F5 (id: ${f5})`,
);
console.log("\nListening for hotkeys... Press Ctrl+C to exit.");

// Using the background-thread callback system
mgr.listen((err, event) => {
  if (err) {
    console.error("Hotkey error:", err);
    return;
  }

  let name = "Unknown";
  if (event.id === shiftM) name = "Ctrl+Alt+Shift+M";
  else if (event.id === f5) name = "F6";

  console.log(`Hotkey event triggered: ${name} -> ${event.state}`);

  if (event.state === "pressed") {
    setTimeout(() => {
      sim.typeText(`You pressed ${name}!\n`);
      sim.mouseMove(10, 10, "relative");
    }, 100);
  }
});

// in some conditions, we keep the script running indefinitely:
// setInterval(() => {}, 1000 * 60 * 60);
